### PR TITLE
Add linking options

### PR DESCRIPTION
**Pull request by adriendelsalle**
Description
---

Details:
- add `always_copy` link feature
- add `always_softlink` link feature
- add `allow_softlinks` link fallback
- add flags in umamba CLI and set options properly

### CODE REVIEW

File: src/link.cpp [hunk 1]
```diff
@@ -898,10 +898,44 @@ namespace mamba
...
+                fs::create_hard_link(src, dst, ec);
+
+                if (ec)
+                {
+                    softlink = m_context->allow_softlinks;
+                    copy = !softlink;
+                }
+                else
+                {
+                    LOG_INFO << "hard linked " << src << " --> " << dst;
```
Comment (wolfv) [inline src/link.cpp:918]: return here? otherwise it can _also_ softlink?

Comment (wolfv) [reply src/link.cpp:918]: ok i think i get it now, looks fine!

File: src/micromamba/main.cpp [hunk 2]
```diff
@@ -915,6 +927,12 @@ init_install_global_parser(CLI::App* subcom)
         ->allow_extra_args(false);
     subcom->add_flag(
         "--no-pin", create_options.no_pin, "Ignore pinned packages (from config or prefix file)");
+    subcom->add_flag(
+        "--allow_softlinks", create_options.allow_softlinks, "Allow softlinks when hardlinks fail");
```
Comment (wolfv) [inline src/micromamba/main.cpp:931]: I prefer `--allow-softlinks` vs underscore

File: include/mamba/transaction_context.hpp [hunk 3]
```diff
@@ -25,14 +26,17 @@ namespace mamba
...
-        TransactionContext(const fs::path& prefix, const std::string& py_version);
+        TransactionContext(Context& ctx, const fs::path& prefix, const std::string& py_version);
 
         bool has_python;
         fs::path target_prefix;
         fs::path site_packages_path;
         fs::path python_path;
         std::string python_version;
         std::string short_python_version;
+        bool allow_softlinks = false;
```
Comment (wolfv) [inline include/mamba/transaction_context.hpp:37]: is this used?

**Review by wolfv — APPROVED** [review]
looks awesome!

File: src/link.cpp (hunk 1 above)
Comment (adriendelsalle) [reply src/link.cpp:918]: cool thanks for the review!

File: src/micromamba/main.cpp (hunk 2 above)
Comment (adriendelsalle) [reply src/micromamba/main.cpp:931]: it's a typo sorry, will be fixed!

File: include/mamba/transaction_context.hpp (hunk 3 above)
Comment (adriendelsalle) [reply include/mamba/transaction_context.hpp:37]: Yes it is:

When a hardlink fails, we have this fallback :
```
softlink = m_context->allow_softlinks;
copy = !softlink;
```

File: src/link.cpp [hunk 4]
```diff
@@ -898,10 +898,44 @@ namespace mamba
             return std::make_tuple(validate::sha256sum(dst), rel_dst);
         }
 
-        if (path_data.path_type == PathType::HARDLINK)
+        if (path_data.no_link || (path_data.path_type == PathType::HARDLINK))
```
Comment (wolfv) [inline src/link.cpp:901]: maybe change the order of comparison here because path_data.path_type == HARDLINK is much more likely (speed boost!)

Comment (adriendelsalle) [reply src/link.cpp]: done ! huge boost ! :)

**Pull request merged**